src/types.rs

```diff
@@ -330,3 +330,134 @@ impl DestinationSlot {
         self.dest.get()
     }
 }
+
+/// Helper structure for parsing the result of a multishot [`opcode::RecvMsg`](crate::opcode::RecvMsg).
+#[derive(Debug)]
+pub struct RecvMsgOut<'buf> {
+    buf: &'buf [u8],
+    header: sys::io_uring_recvmsg_out,
+    /// The fixed length of the name field, in bytes.
+    ///
+    /// If the incoming name data is larger than this, it gets truncated to this.
+    /// If it is smaller, it gets 0-padded to fill the whole field. In either case,
+    /// this fixed amount of space is reserved in the result buffer.
+    msghdr_name_len: usize,
+    /// The fixed length of the control field, in bytes.
+    ///
+    /// This follows the same semantics as the field above, but for control data.
+    msghdr_control_len: usize,
+}
+
+impl<'buf> RecvMsgOut<'buf> {
+    const DATA_START: usize = std::mem::size_of::<sys::io_uring_recvmsg_out>();
+
+    /// Parse the data buffered upon completion of a `RecvMsg` multishot operation.
+    ///
+    /// `buffer` is the whole buffer previously provided to the ring, while `msghdr`
+    /// is the same content provided as input to the corresponding SQE
+    /// (only `msg_namelen` and `msg_controllen` fields are relevant).
+    pub fn parse(buffer: &'buf [u8], msghdr: &libc::msghdr) -> Result<Self, ()> {
+        if buffer.len() < std::mem::size_of::<sys::io_uring_recvmsg_out>() {
+            return Err(());
+        }
+        // SAFETY: buffer (minimum) length is checked here above.
+        let header: sys::io_uring_recvmsg_out =
+            unsafe { std::ptr::read_unaligned(buffer.as_ptr() as _) };
+
+        let msghdr_name_len = msghdr.msg_namelen as _;
```

Comment:
It's better to use `.into()` whenever possible

Reply:
This is a `u32 -> usize` conversion, which is always safe to perform on Linux but unfortunately not yet covered in Rust by `Into` (see https://github.com/rust-lang/rust/issues/70460).

Additionally there are some mismatches across different libc, so that the source fields are `u32` on some libc and `usize` on other libc (so actually not really casting). This is all very unfortunate, but `as _` is pretty much a required hack to appease rustc and clippy in a portable way across libc's.